include/core/CWordDictionary.h

```diff
@@ -72,34 +72,71 @@ class CORE_EXPORT CWordDictionary : private CNonCopyable {
     template<size_t DEFAULT_EXTRA_WEIGHT>
     class CWeightAll {
     public:
-        size_t operator()(EPartOfSpeech partOfSpeech) {
+        std::size_t operator()(EPartOfSpeech partOfSpeech) {
             return (partOfSpeech == E_NotInDictionary) ? 0 : DEFAULT_EXTRA_WEIGHT;
         }
+
+        void reset() {
+            // NO-OP
+        }
     };
 
     using TWeightAll2 = CWeightAll<2>;
 
     //! Functor for weighting one type of dictionary word by a certain
     //! amount and all dictionary words by a different amount
-    template<EPartOfSpeech SPECIAL_PART1, size_t EXTRA_WEIGHT1, size_t DEFAULT_EXTRA_WEIGHT>
+    template<EPartOfSpeech SPECIAL_PART1, std::size_t EXTRA_WEIGHT1, std::size_t DEFAULT_EXTRA_WEIGHT>
     class CWeightOnePart {
     public:
-        size_t operator()(EPartOfSpeech partOfSpeech) {
+        std::size_t operator()(EPartOfSpeech partOfSpeech) {
             if (partOfSpeech == E_NotInDictionary) {
                 return 0;
             }
             return (partOfSpeech == SPECIAL_PART1) ? EXTRA_WEIGHT1 : DEFAULT_EXTRA_WEIGHT;
         }
+
+        void reset() {
+            // NO-OP
+        }
     };
 
     using TWeightVerbs5Other2 = CWeightOnePart<E_Verb, 5, 2>;
 
+    //! Functor for weighting one type of dictionary word by a certain
+    //! amount and all dictionary words by a different amount, giving a boost to
+    //! the weight when 3 or more dictionary words are adjacent one another.
+    template<EPartOfSpeech SPECIAL_PART1, std::size_t EXTRA_WEIGHT1, std::size_t DEFAULT_EXTRA_WEIGHT, std::size_t ADJACENT_PARTS_BOOST>
+    class CWeightOnePartAdjacentBoost {
+    public:
+        std::size_t operator()(EPartOfSpeech partOfSpeech) {
+            if (partOfSpeech == E_NotInDictionary) {
+                m_NumOfAdjacentDictionaryWords = 0;
+                return 0;
+            }
+
+            std::size_t weight = (partOfSpeech == SPECIAL_PART1) ? EXTRA_WEIGHT1 : DEFAULT_EXTRA_WEIGHT;
+            std::size_t boost = (m_NumOfAdjacentDictionaryWords > 1 ? ADJACENT_PARTS_BOOST : 1);
+            weight *= boost;
+
+            m_NumOfAdjacentDictionaryWords++;
```

Comment:
nit: throughout the C++ codebase we use prefix increment when the original value is not used.

```suggestion
            ++m_NumOfAdjacentDictionaryWords;
```